File: 03003. Maximize the Number of Partitions After Operations/solution.go
/**
 * Problem: 3003. Maximize the Number of Partitions After Operations
 *
 * Difficulty: Hard
 *
 * Language: Golang
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

// Data struct represents the state of a segment with its mask, segment count,
// and flags for full prefix and suffix.
type Data struct {
	mask     uint32
	segs     uint16
	fullpref bool
	fullsuff bool
}

// DP array is used to store intermediate results for each index in the string.
var DP [10000]Data

// maxPartitionsAfterOperations calculates the maximum number of partitions
// after performing operations on the string `s` with a limit of `k` unique letters.
func maxPartitionsAfterOperations(s string, k int) int {
	// Get the length of the input string.
	n := len(s)

	// If the string length is 1 or `k` is 26, return 1 as no partitioning is needed.
	if n == 1 || k == 26 {
		return 1
	}

	// Initialize a bitmask to track unique letters in the string.
	var uniqueletters uint32

	// Iterate through the string to populate the unique letters bitmask.
	for i := range s {
		uniqueletters |= 1 << (s[i] - 'a')
	}

	// If the number of unique letters is less than `k`, return 1.
	if bits.OnesCount32(uniqueletters) < k {
		return 1
	}

	// Initialize the first and last elements of the DP array.
	DP[0], DP[n-1] = Data{}, Data{}

	// Variables to track the current mask, segment count, and unique letter count.
	var mask uint32
	var segments uint16
	var count int

	// Iterate through the string to calculate prefix segments.
	for i := range n - 1 {
		// Get the current character's index in the alphabet.
		c := s[i] - 'a'

		// Check if the current character is a new unique letter.
		newbit := 1 &^ int(mask>>c)
		count += newbit

		// Update the mask with the current character.
		bit := uint32(1 << c)
		mask |= bit

		// If the count exceeds `k`, start a new segment.
		if count > k {
			segments++
			count = 1
			mask = bit
		}

		// Update the DP array with the current state.
		DP[i+1] = Data{mask, segments, count == k, false}
	}

	// Reset variables for suffix calculation.
	mask, segments, count = 0, 0, 0

	// Iterate through the string in reverse to calculate suffix segments.
	for i := n - 1; i > 0; i-- {
		// Get the current character's index in the alphabet.
		c := s[i] - 'a'

		// Check if the current character is a new unique letter.
		newbit := 1 &^ int(mask>>c)
		count += newbit

		// Update the mask with the current character.
		bit := uint32(1 << c)
		mask |= bit

		// If the count exceeds `k`, start a new segment.
		if count > k {
			segments++
			count = 1
			mask = bit
		}

		// Update the DP array with the current state.
		DP[i-1].mask |= mask
		DP[i-1].segs += segments
		DP[i-1].fullsuff = count == k
	}

	// Variable to store the maximum number of partitions.
	var result uint16

	// Constant representing the full letter mask (all 26 letters).
	const LETTER_MASK = 1<<26 - 1

	// Iterate through the DP array to calculate the result.
	for i := range n {
		// Check if the current segment has a full prefix and suffix.
		full := DP[i].fullpref && DP[i].fullsuff

		// Get the current mask and segment count.
		mask, seg := DP[i].mask, DP[i].segs

		// Adjust the segment count based on the mask and full flag.
		if full && mask != LETTER_MASK {
			seg += 2
		} else if bits.OnesCount32(mask) >= k {
			seg += 1
		}

		// Update the result with the maximum segment count.
		result = max(result, seg+1)
	}

	// Return the result as an integer.
	return int(result)
}

File: 03003. Maximize the Number of Partitions After Operations/solution.js
/**
 * Problem: 3003. Maximize the Number of Partitions After Operations
 *
 * Difficulty: Hard
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 4 ms (Beats 100%)
 */

/**
 * Max partitions of string s with up to k distinct letters per operation
 *
 * @param {string} s - Input string
 * @param {number} k - Max distinct letters per operation
 *
 * @returns {number} - Max partitions
 */
const maxPartitionsAfterOperations = (s, k) => {
  // If k is 26 (all letters in the alphabet), only one partition is needed
  if (k === 26) return 1

  // Get the length of the input string
  const n = s.length

  // Add sentinel |characters to the string for easier boundary handling
  s = '@' + s + '@'

  // Initialize prefix arrays to store partition counts and bit values
  const pref = new Array(n + 2).fill(0),
    pval = new Array(n + 2).fill(0)

  // Initialize variables for prefix partitioning
  let prefix = 0,
    pbit = 0

  // Iterate through the string to calculate prefix partitions
  for (let i = 1; i <= n; i++) {
    // Calculate the bit representation of the current character
    const bit = 1 << (s.charCodeAt(i) - 'a'.charCodeAt(0))

    // Add the current character's bit to the prefix bitmask
    pbit |= bit

    // If the number of distinct letters exceeds k, start a new partition
    if (countBits(pbit) > k) {
      prefix++
      pbit = bit
    }

    // Store the current prefix partition count and bitmask
    pref[i] = prefix
    pval[i] = pbit
  }

  // Initialize suffix arrays to store partition counts and bit values
  const suff = new Array(n + 2).fill(0),
    sval = new Array(n + 2).fill(0)

  // Initialize variables for suffix partitioning
  let suffix = 0,
    sbit = 0

  // Iterate through the string in reverse to calculate suffix partitions
  for (let i = n; i >= 1; i--) {
    // Calculate the bit representation of the current character
    const bit = 1 << (s.charCodeAt(i) - 'a'.charCodeAt(0))

    // Add the current character's bit to the suffix bitmask
    sbit |= bit

    // If the number of distinct letters exceeds k, start a new partition
    if (countBits(sbit) > k) {
      suffix++
      sbit = bit
    }

    // Store the current suffix partition count and bitmask
    suff[i] = suffix
    sval[i] = sbit
  }

  // Initialize the variable to store the maximum number of partitions
  let ans = 0

  // Iterate through the string to calculate the maximum partitions
  for (let i = 1; i <= n; i++) {
    // Calculate the sum of prefix and suffix partitions
    let val = pref[i - 1] + suff[i + 1]

    // Retrieve the prefix and suffix bitmasks
    const p = pval[i - 1],
      s = sval[i + 1]

    // Combine the prefix and suffix bitmasks
    const x = p | s

    // Adjust the partition count based on the combined bitmask
    if (countBits(x) + 1 <= k) val += 1
    else if (countBits(p) === k && countBits(s) === k && countBits(x) < 26)
      val += 3
    else val += 2

    // Update the maximum partition count
    ans = Math.max(val, ans)
  }

  // Return the maximum number of partitions
  return ans
}

/**
 * Count set bits (1s) in a number's binary representation
 *
 * @param {number} n - Input number
 *
 * @returns {number} - Count of set bits
 */
const countBits = (n) => {
  // Initialize a variable to keep track of the count of set bits
  let count = 0

  // Loop until the number becomes 0
  while (n) {
    // Increment count if the least significant bit is 1
    count += n & 1

    // Right shift the number by 1 to process the next bit
    n >>= 1
  }

  // Return the total count of set bits
  return count
}
